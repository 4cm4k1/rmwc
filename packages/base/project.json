{
  "name": "base",
  "$schema": "../../node_modules/nx/schemas/project-schema.json",
  "sourceRoot": "packages/base/src",
  "projectType": "library",
  "tags": ["publishable", "type:ui-component", "readme"],
  "targets": {
    "lint": {
      "executor": "@nx/eslint:lint",
      "outputs": ["{options.outputFile}"]
    },
    "build": {
      "executor": "@nx/vite:build",
      "dependsOn": [
        {
          "projects": "readme",
          "target": "generate-markdown"
        }
      ],
      "outputs": ["{options.outputPath}"],
      "defaultConfiguration": "production",
      "options": {
        "outputPath": "dist/packages/base"
      },
      "configurations": {
        "development": {
          "mode": "development"
        },
        "production": {
          "mode": "production"
        }
      }
    },
    "test": {
      "executor": "@nx/vite:test",
      "outputs": ["{options.reportsDirectory}"],
      "options": {
        "passWithNoTests": true,
        "reportsDirectory": "../../coverage/packages/base"
      }
    }
  }
}
