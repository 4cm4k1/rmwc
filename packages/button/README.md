# Buttons

Buttons communicate the action that will occur when the user touches them.

- Module **@rmwc/button**
- Import styles:
  - Using CSS Loader
    - import '@rmwc/button/styles';
  - Or include stylesheets
    - **'@material/button/dist/mdc.button.css'**
    - **'@rmwc/@rmwc/icon/icon.css'**
    - **'@material/ripple/dist/mdc.ripple.css'**
- MDC Docs: [https://material.io/develop/web/components/buttons/](https://material.io/develop/web/components/buttons/)

```jsx
<Button label="Button" />
```

```jsx
<>
  <Button label="Icon" icon="favorite" />
  <Button label="Trailing" trailingIcon="keyboard_arrow_right" />
  <Button label="Loading" icon={<CircularProgress />} />
</>
```

```jsx
<>
  <Button label="Raised" raised />
  <Button label="Unelevated" unelevated />
  <Button label="Outlined" outlined />
  <Button label="Dense" dense />
  <Button label="No Ripple" ripple={false} />
</>
```

```jsx
<>
  <Button label="Danger" danger raised />
  <Button label="Danger" danger outlined />
  <Button label="Danger" danger />
</>
```

```jsx
<>
  <Button label="With Theme" raised theme={['secondaryBg', 'onSecondary']} />
  {/**
    This example uses "accent" to control the color of the Ripple.
    See the documentation on Ripples.
  */}
  <Button label="With Theme" theme="secondary" />
</>
```

```jsx
<Button>
  {/** Alternatively pass content as children */}
  As Children
</Button>
```

```jsx
<>
  {/** Wrapping a button in TouchTargetWrapper will automatically set its `touch` prop to true. */}
  <TouchTargetWrapper>
    <Button>Touch Accessible</Button>
  </TouchTargetWrapper>
</>
```

## Button

The Button component.

### Props

| Name           | Type          | Description                                                                                                     |
| -------------- | ------------- | --------------------------------------------------------------------------------------------------------------- |
| `children`     | `ReactNode`   | Content specified as children.                                                                                  |
| `danger`       | `boolean`     | Used to indicate a dangerous action.                                                                            |
| `dense`        | `boolean`     | Make the Button dense.                                                                                          |
| `disabled`     | `boolean`     | Make the button disabled                                                                                        |
| `icon`         | `IconPropT`   | An Icon for the Button                                                                                          |
| `label`        | `any`         | Content specified as a label prop.                                                                              |
| `outlined`     | `boolean`     | Make the button outlined.                                                                                       |
| `raised`       | `boolean`     | Make the Button raised.                                                                                         |
| `ripple`       | `RipplePropT` | Adds a ripple effect to the component                                                                           |
| `touch`        | `boolean`     | Makes the button more touch friendly. This will automatically be set true if used inside of TouchTargetWrapper. |
| `trailingIcon` | `IconPropT`   | A trailing icon for the Button                                                                                  |
| `unelevated`   | `boolean`     | Make the button unelevated.                                                                                     |
