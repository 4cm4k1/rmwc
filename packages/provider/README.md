# Provider

A component that provides global configuration for RMWC.

- Module **@rmwc/provider**

You can wrap your top level App component with RMWC provider to set global configuration options. Just pass the options in as props.

```jsx
`
  import React from 'react';
  import * as ReactDOM from 'react-dom';
  import App from './App'; // your main app component
  import { RMWCProvider } from '@rmwc/provider';

  // This example disables ripples globally by default
  ReactDOM.render(
    <RMWCProvider
      // Globally disable ripples
      ripple={false}
      // Global options for icons
      // Takes the same options as the icon component
      icon={{
        basename: 'material-icons'
      }}
      // Global options for typography
      // allows mapping of a defaultTag or specific classes
      // See the Typography docs for more info
      typography={{
        defaultTag: 'div',
        headline1: 'h1'
      }}
      // Global options for tooltips
      // Takes most of the options for tooltips
      // See the Tooltip docs for more info
      tooltip={{
        align: 'right'
      }}
    >
      <App />
    </RMWCProvider>,
    document.getElementById('root'),
  );
`;
```

```jsx
<RMWCProvider ripple={false}>
  <Button>Click me</Button>
</RMWCProvider>
```

```jsx
<RMWCProvider ripple={true}>
  <Button>Click me</Button>
</RMWCProvider>
```

## RMWCProvider

A provider for setting global options in RMWC.

### Props

| Name         | Type                         | Description                                   |
| ------------ | ---------------------------- | --------------------------------------------- |
| `children`   | `ReactNode`                  | Children to render                            |
| `icon`       | `Partial<IconOptions>`       | Global options for icons                      |
| `ripple`     | `boolean`                    | Enable / Disable interaction ripples globally |
| `tooltip`    | `Partial<TooltipOptions>`    | Global tooltip options                        |
| `typography` | `Partial<TypographyOptions>` | Global typography options                     |
