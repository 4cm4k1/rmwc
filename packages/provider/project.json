{
  "name": "provider",
  "$schema": "../../node_modules/nx/schemas/project-schema.json",
  "sourceRoot": "packages/provider/src",
  "projectType": "library",
  "tags": ["publishable", "type:ui-component", "readme"],
  "targets": {
    "lint": {
      "executor": "@nx/eslint:lint",
      "outputs": ["{options.outputFile}"]
    },
    "build": {
      "executor": "@nx/vite:build",
      "dependsOn": [
        {
          "projects": "readme",
          "target": "generate-markdown"
        }
      ],
      "outputs": ["{options.outputPath}"],
      "defaultConfiguration": "production",
      "options": {
        "outputPath": "dist/packages/provider"
      },
      "configurations": {
        "development": {
          "mode": "development"
        },
        "production": {
          "mode": "production"
        }
      }
    },
    "test": {
      "executor": "@nx/vite:test",
      "outputs": ["{options.reportsDirectory}"],
      "options": {
        "passWithNoTests": true,
        "reportsDirectory": "../../coverage/packages/provider"
      }
    },
    "npm-deploy": {
      "executor": "ngx-deploy-npm:deploy",
      "options": {
        "access": "public",
        "distFolderPath": "dist/packages/provider"
      },
      "dependsOn": ["pre-npm-deploy-build-production"]
    },
    "pre-npm-deploy-build-production": {
      "executor": "nx:run-commands",
      "options": {
        "command": "nx run provider:build:production"
      }
    }
  }
}
