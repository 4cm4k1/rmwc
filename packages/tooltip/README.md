# Tooltips

Tooltips display informative text when users hover over, focus on, or tap an element.

- Module **@rmwc/tooltip**
- Import styles:
  - Using CSS Loader
    - import '@rmwc/tooltip/styles';
  - Or include stylesheets
    - **'@material/tooltip/dist/mdc.tooltip.css'**
    - **'@rmwc/tooltip/tooltip.css'**

## Basic Usage

Wrap any component in a `Tooltip` and provide the overlay attribute. The only requirement is that is has a single React child, and that the child can accept `onMouseEnter`, `onMouseLeave`, `onFocus`, and `onClick` props. This component is not backwards compatible with the tooltip from RMWC version 8.x. For a compatible version us the `RCTooltip` component. The `RCTooltip` uses the `ReactTooltip` from 'rc-tooltip' package. This tooltip uses the standard tooltip component from Google Material for the web.

```jsx
<>
  <Tooltip overlay="Cookies">
    <IconButton icon="star_border" aria-describedby="tooltip-id" />
  </Tooltip>

  <Tooltip overlay="Pizza">
    <IconButton icon="favorite_border" />
  </Tooltip>

  <Tooltip overlay="Icecream">
    <IconButton icon="mood" />
  </Tooltip>
</>
```

## Variants

```jsx
<Tooltip overlay="Hello" open={true}>
  <IconButton icon="mood" />
</Tooltip>
```

```jsx
<Tooltip
  overlay={
    <>
      <RichTooltipTitle>Hello</RichTooltipTitle>
      <RichTooltipContent>
        I am the content of the interactive rich tooltip
      </RichTooltipContent>
      <RichTooltipActions>
        <Button>Click me</Button>
      </RichTooltipActions>
    </>
  }
>
  <span role="button">
    Popover with clickable content that stays open on hover
  </span>
</Tooltip>
```

```jsx
<Tooltip
  overlay={
    <SimpleRichTooltip
      title="My title"
      body="I am the content"
      actions={<Button>Click me</Button>}
    />
  }
>
  <span role="button">Usage of SimpleRichTooltip</span>
</Tooltip>
```

```jsx
<Tooltip
  overlay={
    <SimpleRichTooltip
      title="My title"
      body={
        <RichTooltipLink href="/" target="_blank">
          Link
        </RichTooltipLink>
      }
    />
  }
>
  <span role="button">With links</span>
</Tooltip>
```

```jsx
<Tooltip
  overlay={<RichTooltipContent>I am persistent</RichTooltipContent>}
  isPersistent
>
  <span role="button">Popover when I am clicked</span>
</Tooltip>
```

```jsx
<Tooltip
  overlay={<RichTooltipContent>I don't stay open</RichTooltipContent>}
  stayOpenOnHover={false}
>
  <span role="button">I don't stay open</span>
</Tooltip>
```

```jsx
<Tooltip
  rich={false}
  overlay={
    <div
      style={{
        display: 'flex',
        alignItems: 'center',
        justifyContent: 'center',
        background: 'white',
        width: '20rem',
        height: '8rem',
        color: 'black',
        borderRadius: '3px',
        margin: '0 -3px'
      }}
    >
      Hello world
    </div>
  }
>
  <span role="button">Popover Window</span>
</Tooltip>
```

```jsx
<>
  <Tooltip overlay="Cookies" enterDelay={1000}>
    <Button label="Enter Delay" />
  </Tooltip>

  <Tooltip overlay="Pizza" leaveDelay={1000}>
    <Button label="Leave Delay" />
  </Tooltip>

  <Tooltip overlay="Icecream" enterDelay={1000} leaveDelay={1000}>
    <Button label="Both" />
  </Tooltip>
</>
```

```jsx
<>
  <Tooltip overlay="Align start" align="start">
    <IconButton icon="trip_origin" />
  </Tooltip>
  <Tooltip overlay="Align center" align="center">
    <IconButton icon="trip_origin" />
  </Tooltip>
  <Tooltip overlay="Align end" align="end">
    <IconButton icon="trip_origin" />
  </Tooltip>
  <Tooltip overlay="Align above" align="above">
    <IconButton icon="trip_origin" />
  </Tooltip>
  <Tooltip overlay="Align below" align="below">
    <IconButton icon="trip_origin" />
  </Tooltip>
</>
```

## Activation

By default, tooltips will activate on hover and focus. You can change this behavior by passing one or more options to the `activateOn` prop.

```jsx
<>
  <Tooltip overlay="Cookies" activateOn="hover">
    <Button label="Hover" />
  </Tooltip>
  <Tooltip overlay="Pizza" activateOn="click" isPersistent>
    <Button label="Click" />
  </Tooltip>
  <Tooltip overlay="Icecream" activateOn="focus">
    <Button label="Focus" />
  </Tooltip>
  <Tooltip overlay="Cake" activateOn={['hover', 'focus']}>
    <Button label="Multiple" />
  </Tooltip>
</>
```

## Usage with RMWCProvider

The RMWCProvider allows you to specify global defaults for your tooltips.

```jsx
<RMWCProvider
  tooltip={{
    align: 'right',
    activateOn: 'hover',
    leaveDelay: 500,
    enterDelay: 0
  }}
>
  <Tooltip overlay="Hello World!">
    <Button label="With Provider" />
  </Tooltip>
</RMWCProvider>
```

## Tooltip

A Tooltip component for displaying informative popover information.

### Props

| Name         | Type                                         | Description                                                        |
| ------------ | -------------------------------------------- | ------------------------------------------------------------------ |
| `activateOn` | `TooltipActivationT \| TooltipActivationT[]` | Activate the tooltip through one or more interactions. Defaults to |

`['hover', 'focus']`
. |
| `align` | `TooltipAlignT` | How to align the tooltip. |
| `anchorBoundaryType` | `AnchorBoundaryType` | Specify whether the anchor element is bounded (element has an identifiable boundary such as a button) or unbounded (element does not have a visually declared boundary such as a text link). |
| `children` | `ReactNode` | The children that the tooltip belongs to. Must be a single React element. |
| `className` | `string` | Custom className to add to the tooltip overlay container. |
| `enterDelay` | `number` | Delay in milliseconds before showing the tooltip when interacting via touch or mouse. |
| `isPersistent` | `boolean` | Specify whether tooltip should be persistent. Persistent tooltip are triggered by clicks. |
| `leaveDelay` | `number` | Delay in milliseconds before hiding the tooltip when interacting via touch or mouse. |
| `open` | `boolean` | Manually control the open state |
| `overlay` | `ReactNode` | The overlay for the tooltip. |
| `rich` | `boolean` | Manually disable a tooltip being rich. This is useful when a custom overlay is needed. Defaults to true when content is a ReactNode. |
| `stayOpenOnHover` | `boolean` | Control whether to stay open on hover. This is useful for interactive tooltips. |
