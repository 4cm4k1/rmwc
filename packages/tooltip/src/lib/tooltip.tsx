import React, { ReactNode } from 'react';
import * as RMWC from '@rmwc/types';
import {
  classNames,
  createComponent,
  PortalChild,
  PortalPropT,
  Tag,
  useId
} from '@rmwc/base';
import { useToolTipFoundation } from './foundation';
import { AnchorBoundaryType } from '@material/tooltip';
import { TooltipAlignT } from '@rmwc/provider';

/** A Tooltip component for displaying informative popover information. */
export interface TooltipProps {
  /** How to align the tooltip. */
  align?: TooltipAlignT;
  /** The overlay for the tooltip. */
  overlay?: React.ReactNode;
  /** The children that the tooltip belongs to. Must be a single React element. */
  children: React.ReactNode;
  /** Custom className to add to the tooltip overlay container. */
  className?: string;
  /** Delay in milliseconds before showing the tooltip when interacting via touch or mouse. */
  enterDelay?: number;
  /** Delay in milliseconds before hiding the tooltip when interacting via touch or mouse. */
  leaveDelay?: number;
  /** Specify whether the anchor element is bounded (element has an identifiable boundary such as a button) or unbounded (element does not have a visually declared boundary such as a text link). */
  anchorBoundaryType?: AnchorBoundaryType;
  /** Renders the tooltip to a portal. Defaults to true. */
  renderToPortal?: PortalPropT;
}

export type TooltipHTMLProps = RMWC.HTMLProps<HTMLDivElement>;

/** A Tooltip component for displaying informative popover information. */
export const Tooltip: RMWC.ComponentType<
  TooltipProps,
  TooltipHTMLProps,
  'div'
> = createComponent<TooltipProps, TooltipHTMLProps>(
  function Tooltip(props, ref) {
    const {
      anchorEl,
      rootEl,
      isShown,
      surfaceEl,
      handleMouseEnter,
      handleMouseLeave,
      handleFocus,
      handleTransitionEnd,
      handleTouchstart,
      handleTouchend
    } = useToolTipFoundation(props);

    const { overlay, renderToPortal = true } = {
      ...props
    };

    const uniqueId = useId('tooltip', props);

    const className = classNames('mdc-tooltip');

    const child = React.Children.only(props.children);

    if (!React.isValidElement<React.HTMLProps<any>>(child)) {
      return null;
    }

    return (
      <>
        {React.cloneElement(child, {
          ...anchorEl.props(child.props),
          element: anchorEl,
          ref: anchorEl.reactRef,
          'aria-describedby': uniqueId,
          'data-tooltip-id': uniqueId,
          onMouseEnter: handleMouseEnter,
          onMouseLeave: handleMouseLeave,
          onFocus: handleFocus,
          onTouchStart: handleTouchstart,
          onTouchEnd: handleTouchend
        })}
        <PortalChild renderTo={renderToPortal}>
          <Tag
            tag="div"
            className={className}
            id={uniqueId}
            role="tooltip"
            aria-hidden={isShown}
            ref={ref}
            element={rootEl}
            onTransitionEnd={handleTransitionEnd}
          >
            <Tag
              tag="div"
              element={surfaceEl}
              className="mdc-tooltip__surface mdc-tooltip__surface-animation"
            >
              {overlay}
            </Tag>
          </Tag>
        </PortalChild>
      </>
    );
  }
);

export interface RichTooltipTitleProps {}

export const RichTooltipTitle = createComponent<RichTooltipTitleProps>(
  function RichTooltipTitle(props, ref) {
    const { children } = props;
    return (
      <Tag tag="h2" className="mdc-tooltip__title" {...props} ref={ref}>
        {children}
      </Tag>
    );
  }
);

export interface RichTooltipContentProps {}

export const RichTooltipContent = createComponent<RichTooltipContentProps>(
  function RichTooltipContent(props, ref) {
    const { children, ...rest } = props;
    return (
      <Tag tag="p" className="mdc-tooltip__content" {...rest} ref={ref}>
        {children}
      </Tag>
    );
  }
);

export interface RichTooltipActionsProps {}

export const RichTooltipActions = createComponent<RichTooltipActionsProps>(
  function RichTooltipActions(props, ref) {
    const { children, ...rest } = props;
    return (
      <Tag tag="p" className="mdc-tooltip--rich-actions" {...rest} ref={ref}>
        {children}
      </Tag>
    );
  }
);

export const RichTooltipLink = createComponent(
  function RichTooltipLink(props, ref) {
    const { href, children, ...rest } = props;
    return (
      <Tag
        tag="a"
        className="mdc-tooltip__content-link"
        href={href}
        {...rest}
        ref={ref}
      >
        {children}
      </Tag>
    );
  }
);

export interface RichTooltipProps extends TooltipProps {
  title?: ReactNode;
  body?: ReactNode;
  actions?: ReactNode;
  link?: ReactNode;
  isPersistent?: boolean;
}

export const RichTooltip: RMWC.ComponentType<
  RichTooltipProps,
  TooltipHTMLProps,
  'div'
> = createComponent<RichTooltipProps, TooltipHTMLProps>(
  function RichTooltip(props, ref) {
    const {
      anchorEl,
      rootEl,
      isShown,
      surfaceEl,
      handleMouseEnter,
      handleMouseLeave,
      handleFocus,
      handleTransitionEnd,
      handleClick,
      handleTouchstart,
      handleTouchend
    } = useToolTipFoundation(props);

    const { title, body, actions, link, isPersistent, overlay } = {
      ...props
    };

    const uniqueId = useId('tooltip', props);

    const className = classNames('mdc-tooltip', 'mdc-tooltip--rich');

    const eventHandlers = isPersistent
      ? { onClick: handleClick }
      : {
          onMouseEnter: handleMouseEnter,
          onMouseLeave: handleMouseLeave,
          onFocus: handleFocus,
          onTouchStart: handleTouchstart,
          onTouchEnd: handleTouchend
        };

    const child = React.Children.only(props.children);
    if (!React.isValidElement<React.HTMLProps<any>>(child)) {
      return null;
    }

    return (
      <Tag tag="div" className="mdc-tooltip-wrapper--rich">
        {React.cloneElement(child, {
          ...anchorEl.props(child.props),
          element: anchorEl,
          ref: anchorEl.reactRef,
          'aria-describedby': uniqueId,
          'aria-haspopup': 'dialog',
          'data-tooltip-id': uniqueId,
          ...eventHandlers
        })}
        <Tag
          tag="div"
          className={className}
          id={uniqueId}
          role={actions ? 'dialog' : 'tooltip'}
          aria-hidden={isShown}
          ref={ref}
          element={rootEl}
          data-mdc-tooltip-persistent={isPersistent}
          tabIndex={isPersistent && -1}
          onTransitionEnd={handleTransitionEnd}
          onMouseEnter={!isPersistent && handleMouseEnter}
          onMouseLeave={!isPersistent && handleMouseLeave}
        >
          <Tag
            tag="div"
            element={surfaceEl}
            className="mdc-tooltip__surface mdc-tooltip__surface-animation"
          >
            {overlay}
            {title && <RichTooltipTitle>{title}</RichTooltipTitle>}
            <RichTooltipContent>
              {body}
              {link}
            </RichTooltipContent>
            {actions && <RichTooltipActions>{actions}</RichTooltipActions>}
          </Tag>
        </Tag>
      </Tag>
    );
  }
);
