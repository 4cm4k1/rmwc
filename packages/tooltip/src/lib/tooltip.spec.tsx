import { RMWCProvider } from '@rmwc/provider';
import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { RichTooltip, RichTooltipLink, Tooltip } from './tooltip';
import { Button } from '@rmwc/button';
import { Portal } from '@rmwc/base';

describe('Tooltip', () => {
  it('renders', () => {
    const { asFragment } = render(
      <>
        <Tooltip overlay="tooltip" label="test">
          <span>test</span>
        </Tooltip>
        <Portal />
      </>
    );
    userEvent.hover(screen.getByText('test'));
    expect(screen.getByText('tooltip')).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it('className', () => {
    const { asFragment } = render(
      <>
        <Tooltip overlay="tooltip" className="my-custom-classname" label="test">
          <span>test</span>
        </Tooltip>
        <Portal />
      </>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('enterDelay', () => {
    render(
      <>
        <Tooltip overlay="tooltip" enterDelay={1000}>
          <span>test</span>
        </Tooltip>
        <Portal />
      </>
    );
  });

  it('leaveDelay', () => {
    render(
      <>
        <Tooltip overlay="tooltip" leaveDelay={1000}>
          <span>test</span>
        </Tooltip>
        <Portal />
      </>
    );
  });

  it('works with provider', () => {
    render(
      <RMWCProvider tooltip={{}}>
        <Tooltip overlay="tooltip">
          <span>test</span>
        </Tooltip>
        <Portal />
      </RMWCProvider>
    );
  });

  it('accepts allowed align value from provider context', () => {
    vi.spyOn(console, 'warn');
    render(
      <RMWCProvider tooltip={{ align: 'start' }}>
        <Tooltip label="test">
          <span>test</span>
        </Tooltip>
        <Portal />
      </RMWCProvider>
    );
    expect(console.warn).not.toBeCalled();
  });

  it('warn against wrong align value from provider context', () => {
    vi.spyOn(console, 'warn');
    render(
      <RMWCProvider tooltip={{ align: 'bottom' }}>
        <Tooltip label="test">
          <span>test</span>
        </Tooltip>
        <Portal />
      </RMWCProvider>
    );
    expect(console.warn).toBeCalled();
  });
});

describe('RichTooltip', () => {
  it('renders', () => {
    const { asFragment } = render(
      <RichTooltip title="Title" body="This is content" label="test">
        <span>test</span>
      </RichTooltip>
    );
    userEvent.hover(screen.getByText('test'));
    expect(asFragment()).toMatchSnapshot();
  });
  it('can have interactive content', async () => {
    const onClick = vi.fn();

    render(
      <RichTooltip
        title="Title"
        body="This is content"
        actions={<Button onClick={onClick}>Click me</Button>}
      >
        <span>test</span>
      </RichTooltip>
    );

    expect(onClick).not.toHaveBeenCalled();

    userEvent.hover(screen.getByText('test'));
    userEvent.click(screen.getByText('Click me'));

    await waitFor(() => expect(onClick).toHaveBeenCalled());
  });
  it('can have a link', () => {
    const { asFragment } = render(
      <RichTooltip
        title="Title"
        link={
          <RichTooltipLink href="/" target="_blank">
            Link
          </RichTooltipLink>
        }
        label="test"
      >
        <span>test</span>
      </RichTooltip>
    );
    userEvent.hover(screen.getByText('test'));
    expect(asFragment()).toMatchSnapshot();
  });
  it('className', () => {
    const { asFragment } = render(
      <RichTooltip
        className="my-custom-classname"
        title="Title"
        body="This is content"
        label="test"
      >
        <span>test</span>
      </RichTooltip>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
