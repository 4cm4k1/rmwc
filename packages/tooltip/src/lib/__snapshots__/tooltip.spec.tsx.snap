// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`RichTooltip > can have a link 1`] = `
<DocumentFragment>
  <div
    class="mdc-tooltip-wrapper--rich"
  >
    <span
      aria-describedby="tooltip-test"
      aria-haspopup="dialog"
      class=""
      data-tooltip-id="tooltip-test"
      element="[object Object]"
    >
      test
    </span>
    <div
      aria-hidden="false"
      class="mdc-tooltip mdc-tooltip--rich"
      id="tooltip-test"
      role="tooltip"
    >
      <div
        class="mdc-tooltip__surface mdc-tooltip__surface-animation"
      >
        <h2
          class="mdc-tooltip__title"
        >
          Title
        </h2>
        <p
          class="mdc-tooltip__content"
        >
          <a
            class="mdc-tooltip__content-link"
            href="/"
            target="_blank"
          >
            Link
          </a>
        </p>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`RichTooltip > className 1`] = `
<DocumentFragment>
  <div
    class="mdc-tooltip-wrapper--rich"
  >
    <span
      aria-describedby="tooltip-test"
      aria-haspopup="dialog"
      class=""
      data-tooltip-id="tooltip-test"
      element="[object Object]"
    >
      test
    </span>
    <div
      aria-hidden="false"
      class="my-custom-classname mdc-tooltip mdc-tooltip--rich"
      id="tooltip-test"
      role="tooltip"
    >
      <div
        class="mdc-tooltip__surface mdc-tooltip__surface-animation"
      >
        <h2
          class="mdc-tooltip__title"
        >
          Title
        </h2>
        <p
          class="mdc-tooltip__content"
        >
          This is content
        </p>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`RichTooltip > renders 1`] = `
<DocumentFragment>
  <div
    class="mdc-tooltip-wrapper--rich"
  >
    <span
      aria-describedby="tooltip-test"
      aria-haspopup="dialog"
      class=""
      data-tooltip-id="tooltip-test"
      element="[object Object]"
    >
      test
    </span>
    <div
      aria-hidden="false"
      class="mdc-tooltip mdc-tooltip--rich"
      id="tooltip-test"
      role="tooltip"
    >
      <div
        class="mdc-tooltip__surface mdc-tooltip__surface-animation"
      >
        <h2
          class="mdc-tooltip__title"
        >
          Title
        </h2>
        <p
          class="mdc-tooltip__content"
        >
          This is content
        </p>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Tooltip > className 1`] = `
<DocumentFragment>
  <span
    aria-describedby="tooltip-test"
    class=""
    data-tooltip-id="tooltip-test"
    element="[object Object]"
  >
    test
  </span>
  <div
    id="rmwcPortal"
  >
    <div
      aria-hidden="false"
      class="my-custom-classname mdc-tooltip"
      id="tooltip-test"
      role="tooltip"
    >
      <div
        class="mdc-tooltip__surface mdc-tooltip__surface-animation"
      >
        tooltip
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Tooltip > renders 1`] = `
<DocumentFragment>
  <span
    aria-describedby="tooltip-test"
    class=""
    data-tooltip-id="tooltip-test"
    element="[object Object]"
  >
    test
  </span>
  <div
    id="rmwcPortal"
  >
    <div
      aria-hidden="false"
      class="mdc-tooltip"
      id="tooltip-test"
      role="tooltip"
    >
      <div
        class="mdc-tooltip__surface mdc-tooltip__surface-animation"
      >
        tooltip
      </div>
    </div>
  </div>
</DocumentFragment>
`;
